### PR TITLE
Upload Metadata

### PR DESCRIPTION
**Pull request by DwayneJengSage**
See JIRA https://sagebionetworks.jira.com/browse/BRIDGE-1431 and design doc https://docs.google.com/a/sagebase.org/document/d/1dnZ8PaJ7G9H93fBOS7OqSzj2V_MbouGx1nwAaIYa2ZE/edit?usp=sharing

This change adds the upload metadata fields to the Study as well as the Upload API and the Synchronous Health Data API.

Testing done: updated unit tests and integ tests

### CODE REVIEW

**Comment by DwayneJengSage** [comment]
Also added Worker to the get study by ID calls. BridgeEX will need to call Bridge Server to get the study to get upload metadata field defs.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java [hunk 1]
```diff
@@ -203,6 +206,19 @@ public void setUsesCustomExportSchedule(boolean usesCustomExportSchedule) {
     }
 
     /** {@inheritDoc} */
+    @DynamoDBTypeConverted(converter = DynamoUploadSchema.FieldDefinitionListMarshaller.class)
+    @Override
+    public List<UploadFieldDefinition> getUploadMetadataFieldDefinitions() {
+        return uploadMetadataFieldDefinitions;
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:214]: Prevent return of null?

File: app/org/sagebionetworks/bridge/services/StudyService.java [hunk 2]
```diff
@@ -458,6 +466,12 @@ public Study updateStudy(Study study, boolean isAdminUpdate) {
 
         Validate.entityThrowingException(validator, study);
 
+        // Only admins can delete or modify upload metadata fields. Check this after validation, so we don't have to
+        // deal with duplicates.
+        if (!isAdminUpdate) {
+            checkUploadMetadataConstraints(originalStudy, study);
+        }
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/StudyService.java]: But client devs can add?

File: app/org/sagebionetworks/bridge/validators/HealthDataRecordValidator.java [hunk 3]
```diff
@@ -106,6 +106,12 @@ public void validate(Object object, Errors errors) {
                 errors.rejectValue("uploadDate", Validate.CANNOT_BE_NULL);
             }
 
+            // userMetadata is optional, but if specified, it must be an object node.
+            JsonNode userMetadata = record.getUserMetadata();
+            if (userMetadata != null && !userMetadata.isObject()) {
+                errors.rejectValue("userMetadata", "must be an object node");
+            }
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/validators/HealthDataRecordValidator.java:113]: So, do we ever validate that the JSON expresses the FieldDefinitions defined for a study? Shouldn't we? At least the keys of the object should match.

File: app/org/sagebionetworks/bridge/validators/UploadFieldDefinitionListValidator.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+import org.sagebionetworks.bridge.BridgeUtils;
+import org.sagebionetworks.bridge.models.upload.UploadFieldDefinition;
+import org.sagebionetworks.bridge.models.upload.UploadFieldType;
+import org.sagebionetworks.bridge.upload.UploadUtil;
+
+/**
+ * Validates a List of UploadFieldDefinitions. This doesn't use the Validator class, because (1) the typing for a List
+ * of UploadFieldDefinitions is a little wonky and (2) we needed to pass in a little more context to get this to work.
+ */
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/validators/UploadFieldDefinitionListValidator.java:20]: This is fine. I usually pass it in via the constructor, which means we create the validator in a call and then dispose of it.

Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/validators/UploadFieldDefinitionListValidator.java:20]: I prefer the opposite: singletons with no state whose behavior can be determined entirely by their inputs.

File: app/org/sagebionetworks/bridge/validators/HealthDataRecordValidator.java (hunk 3 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/validators/HealthDataRecordValidator.java:113]: No. As it turns out, too much validation makes things too rigid and slows down app development. It's the same reason why most schema fields end up being marked required=false and why StrictValidation, in hindsight, was a mistake.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java (hunk 1 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoStudy.java:214]: Changed to look like the other collections in this class.

File: app/org/sagebionetworks/bridge/services/StudyService.java (hunk 2 above)
Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/services/StudyService.java]: Yes. Added to the comment "Anyone (admin or developer) can add or re-order fields."

**Review by alxdarksage — APPROVED** [review]

**Pull request merged**